### ISSUE TITLE
Advanced upgrade fails table prefix check for new database user

### ISSUE THREAD

**Issue by pixelchutes** [modxcms/revolution#13149]
### Summary
During an "Advanced Upgrade", changing the configured database / user credentials can lead to failure during the table prefix check if the _new_ database/user (SiteB) does not have `SELECT` permission granted to the _old_ (SiteA) MySQL Database.

For example, if you are cloning a previously established MODX site (database + web files), and then running the setup/installer against the new instance.

### Step to reproduce
1. Clone SiteA.com database
2. Clone SiteA.com web files
3. Update all `config.core.php` files
4. (optional) Update `_build/build.config.php` (new db + credentials, if using Git Installation workflow) 
5. (optional) Run `_build/transport.core.php` (for Git)
6. Browse to /setup installer (run "Advanced Setup")
7. Update db + user + password
    - _(**Important**: Different MySQL user, specific to SiteB.com with no access to SiteA.com's database)_
8. Errors on table prefix check

> Checking database: Table prefix does not exist in this database!

### Observed behavior

Error is caught during setup. Cannot proceed with upgrade.

> Checking database: Table prefix does not exist in this database!

This is false, since the new (SiteB) database does exist and with the defined table prefix.

### Expected behavior

If the database or database credentials have changed, and subsequent queries _(especially any validating tables within)_ should **use the new database credentials**. Otherwise, you cannot proceed with the upgrade without [hacking the core](https://github.com/modxcms/revolution/blob/fbf126690f6ce903faeaeb9f1d958211fc2203c9/setup/processors/database/collation.php#L83-L85) or _(temporarily?)_ extending MySQL User Permissions beyond the desired strictness.

### Related
See [slack discussion](http://logs.modx.org/%23general/2016-10-14.log)

### Environment
MODX 2.5.1, MySQL 5.6, PHP 5.6

**Comment by OptimusCrime** [comment]
I investigated this through the related issue #13286 and this one is more or less the same. If the current user lacks permissions to do certain things (select, truncate, delete, alter) MODX will report incorrect error messages instead.

There's no way to get around this as far as I know. You can not hack neither the MODX core to fix this issue, nor simply give the database user more permissions. Instead we should try to figure out if a user has the correct permissions. I do not know exactly how to do this, but it would make it possible to give more correct error messages to the user.

**Comment by pixelchutes** [comment]
This one may be more of an edge case, as setup is using the pre-established xPDO connection for issuing subsequent queries on the advanced setup checks. One approach (for this issue) could be to establish a secondary database connection with the _new_ credentials, or attempting overriding the current (read: previous) connection.

**Comment by OptimusCrime** [comment]
I think you are overcomplicating things, and I do not understand what connections has to do with this? It's just a matter of not having the correct permissios for the current MySQL user to the database? That can be checked, or the error message can be adjusted to include information that this may also be the cause.

Mind elaborating that you mean about the connections some more?

**Comment by pixelchutes** [comment]
Sure. The table prefix check uses the established xPDO connection to check the potentially _new_ database (and username/password for that matter)

It's checking to see if tables already exist, in my case described above, against a new database, because I'm referring to a cloning process vs. straight upgrade.

**Comment by OptimusCrime** [comment]
I see the code you posted now and sure, this is correct, but there is no way to get around this? It can not be done as we lack the propper MySQL permissions?

**Comment by OptimusCrime** [comment]
Also, this would be a problem in many more cases too? Like during a regular installation.

**Comment by pixelchutes** [comment]
As I understand it, I don't think it would impact a new install. Different scenarios based on my understanding of what I think is happening.

**Comment by OptimusCrime** [comment]
You're right, perhaps not the SELECT permission happens, but a regular installation will require CREATE permissions, upgrades also needs ALTER and perhaps DROP. Clear the cache (among other things) needs TRUNCATE etc.

I belive that checking that all of these permissions are given during installation would be a good idea either way, as we can not check this for every single query. My five cents is therefore to include that as a step in the upgrade/setup process.
